### DISCUSSION TITLE
How to Centralize Concurrency Management

### DISCUSSION THREAD

**Discussion by izhouteng** [OptimalBits/bull#2429] [Q&A]
## Description

1 node instance, this code works fine.

But I start 8 instances with PM2 and actually the concurrency will become 40.

How can I control concurrency under multiple instances of PM2?

```js
const bullQueue = require('bull');
const queue = new bullQueue('taskqueue', {...})
const concurrency = 5;
queue.process('jobTypeA', concurrency, job => {...do something...});
```

**Comment by manast** [comment]
I am not sure I understand the question, because just stating the obvious and what you said yourself, you control the number of workers with pm2 and the concurrency per worker with the concurrency option...

**Reply by izhouteng** [reply]
But 1 worker cannot take advantage of multi-core CPU performance.

In fact, I use the nestjs framework, the above code is just to illustrate the problem.

I want to start 8 instances with PM2 and want to set total concurrency for all instances instead of concurrency per instance.

how to set up global concurrency? Just like limiter.

**Reply by manast** [reply]
When you start workers using pm2 you are creating new processes, one per instance, so they are using multiple cores. Of course you can configure every worker to have whatever concurrency factor you want, and that is just determining how many parallel jobs a given worker can run (within one process).

**Reply by manast** [reply]
Not sure if what you are referring to is something like max concurrency for groups: https://docs.bullmq.io/bullmq-pro/groups/concurrency

**Reply by izhouteng** [reply]
Thanks for your answer.

finally I found a solution in this issue. It also applies to bull. and the nestjs framework.

https://github.com/bee-queue/bee-queue/issues/148

I think Bull's Rate limiter can consider supporting the concurrency parameter to control the global concurrency.

This seems to be a much talked about feature. I saw a lot of people on stackoverflow have the same doubts as me.

**Reply by the-spyke** [reply]
@manast Could you please go into more details on how rate limiter works and how it is different to concurrency. I have a similar issue to the OP, but in my case its multiple k8s processing pods for 1 queue. I saw in moveToActive.lua that rate limiter is a part of task data, but what if the task was queued by another Queue instance without any options?

**Reply by manast** [reply]
@the-spyke I just wrote an article about rate-limit, although is for BullMQ, nevertheless if you are starting with Bull I would recommend that you use BullMQ: https://blog.taskforce.sh/rate-limit-recipes-in-nodejs-using-bullmq/

**Reply by the-spyke** [reply]
@manast No, I have an old project using Bull 3.18 and try to solve performance issues when adding 1k of jobs loads Redis CPU to 100% and makes it unavailable.

**Reply by manast** [reply]
@the-spyke Ok, and you have a rate-limit enabled in this case?

**Reply by manast** [reply]
Sounds strange that just 1k would create such effect on the CPU, something else must be at play here.

**Reply by the-spyke** [reply]
@manast I don't have a rate limiter. That's why I started to read the docs about it and concurrency option. After dispatching the batch I see a huge spike in SetTypeCmds (others too, but much less). I'd like to understand how it should work before I will change any code.

For example, concurrency is 1 by default, which means you need rate limiter only if you want to "add pauses" between tasks. And in case of a k8s concurrency will be equal to the scaling factor of the service. Also, how rate limiter would work, if I have 2 workers with
```ts
new Queue('name', { limiter: { max 5, duration: 1000 } });
```
and another service adding tasks into
```
new Queue('name');
```
Another one: how concurrency works if the queue has only one handler registered as `*`.

**Reply by manast** [reply]
Adding 1k jobs (if done in parallel) would take less than one second, so no chance that the Redis CPU would create such a big spike. The concurrency setting just describes how many jobs a given worker (a Queue instance with a process function defined) will process in parallel. For IO based jobs a concurrency of something between 100 and 300 works normally quite well.
For the rest of the questions you should be able to find everything you ask in the documentation.

**Reply by the-spyke** [reply]
My guess was that the spike appears because of too many concurrent workers. But docs don't say anything special about `*`, so in my case it should be default concurrency 1 multiplied by 3 (scaling). So, I decided to look into rate limiting. I suppose every queue instance manages rate limit for itself, but seeing limiter code in the `moveToActive.lua` made me wonder if it is baked into job data by the producing queue instead.
